source/client.c

```diff
@@ -308,18 +333,21 @@ struct aws_mqtt_client_connection *aws_mqtt_client_connection_new(struct aws_mqt
 
     if (aws_mutex_init(&connection->pending_requests.mutex)) {
 
-        goto handle_error;
+        AWS_LOGF_ERROR(AWS_LS_MQTT_CLIENT, "Failed to initialize pending_requests mutex");
+        goto failed_init_pending_requests_mutex;
```

Comment:
Given the ZERO_STRUCT call here, it seems like this could be cleaned up to have a single on_error that destroys the connection, which would simplify the logic a lot.  Your destroy might need to be made a bit more paranoid in spots to make sure it handles the partial construction cases.